### DISCUSSION TITLE
How might Statiq handle "Gated" content?

### DISCUSSION THREAD

**Discussion by jonhilt** [statiqdev/Discussions#4] [Q&A]
Hey, I've only just really started exploring Statiq so apologies if this is covered elsewhere.

I have a use-case I wanted to quickly explore and discuss how/whether Statiq will support it.

As well as my [static blog](https://jonhilton.net) (which is currently created using Hugo) I have just launched a [new site to host my books/courses](https://practicaldotnet.io) etc.

I chose to build this using Blazor Server (with a view to migrating it to Blazor WASM) but I've realised that Blazor is pretty heavy for a lot of the site which is really just landing pages which could easily be static.

Here's an example: [https://practicaldotnet.io/c/BlazorByExample](https://practicaldotnet.io/c/BlazorByExample)

So I think generating these as static pages makes much more sense (all the content is Markdown anyway).

So, to my question.

I'm confident Statiq will handle the landing pages part etc. But what about "gated content".

[This page](https://practicaldotnet.io/c/PracticalMVC/002-list-kanban-boards-part-1) for example shows how you can browse content, but when you attempt to move on to the next module, you would need to purchase the book to continue.

Essentially this is all still static content, but at the moment it's retrieved via an API call the backend which can determine if the visitor has access to the content.

At a high level, how might Statiq support this requirement do we think?

Thanks for the work so far, this project seems to be shaping up very nicely and I'd love to switch my blog from Hugo in the near future!

**Comment by daveaglick** [comment]
Great question!

Stepping back and looking at the more general concept of gated static site content, I can come up with a few different approaches that could apply to any JAMstack site and generator:
- Relying on capabilities of the static host such as [Netlify Identity](https://docs.netlify.com/visitor-access/identity).
- Splitting the site into a set of static landing pages on one host/service and a web server backed site with identity for the gated content, with a proxy like Front Door to unify the two.
- Using an API to authenticate and provide the gated content to the client (which is what it sounds like you're currently doing, and probably also how I'd do this).

So let's take that last approach and look at how Statiq might make it easier. I think the key here is that Statiq is just a .NET application (including [Statiq Web](https://github.com/statiqdev/Statiq.Web) since it's built on top of Statiq Framework). It can be extended through .NET code and there are lots of ways to do so, including plugging into the deployment process. That provides a high degree of flexibility to adapt to any static generation scenario - it's why I like to say Statiq is designed for building _your_ static generator instead of just being a static generator.

More specifically, and off the top of my head, the way I'd approach this is:
- Include the gated content right alongside the non-gated content. This makes managing everything easy. Designate the difference using relative paths or maybe metadata (for example, front matter like `Gated: true`).
- Use a global setting to indicate if you're doing a local build (for example, on the command line like `--setting IncludeGated=true`) and then treat the gated content normally if so (I.e., don't filter it out in the step below) - that way while you're developing it gets included right alongside your normal ungated content.
- Output the non-gated content normally, but filter out the gated content.
- Direct the gated content to a separate pipeline that processes it and then either keeps it in memory (for deployment to an API or service like Functions, Cosmos DB, or Azure Search) or writes it to a different output folder (for deployment as static files to a different location/host).
- Create a separate deployment pipeline for the gated content that deploys it to the appropriate location/API.

There's likely other ways Statiq could handle this scenario too - that's just the first approach that comes to mind. Hopefully it was clear(ish)? I think the more general point is that because it's .NET code driving everything, and you can add new .NET code in the form of [modules](https://statiq.dev/framework/pipelines/modules/) and [pipelines](https://statiq.dev/framework/pipelines/), the generator itself can adapt to more exotic use cases like this.

**Comment by jonhilt** [comment]
Thanks for that, I read it a few times to try and unpack the details (some of the Statiq terminology is still new to me!) but I think I understand in theory at least!

Currently, I'm writing everything in Markdown and serving it via an API (using Markdig to turn the Markdown into HTML), using YAML front matter to indicate whether the content is a FreePreview or not.

So it seems I'm already on a good path to migrate this to Statiq and then just have the same API return the generated content.

I think my next step will be to set up a spike to test this out, taking the existing content and the approach you identified above to pipe the content to two separate locations.

I'd probably want to show consolidated navigation for all the content (free or gated) as I am currently. So I'm assuming the pipeline could spit out nav links for everything (essentially a table of contents) but take the "second pipeline" approach to actually generate and deploy the gated content to a different location.

Thanks again for thinking this through and taking the time to point me in the right direction. Now I have some fun learning to do this weekend :-)

**Comment by jonhilt** [comment]
Hello again!

I'm making progress and have managed to output my landing pages using their existing styles etc. to the default output folder.

So all good to this point 👍 

One thing I can't quite figure out (yet) is how to write a Deployment pipeline as suggested. Specifically, I've figured out I can indicate `Deployment  = true`, but the bit I'm stuck on is coping the output files to another location.

Is it a case of using the `CopyFiles` module as an `OutputModule`?

If so, how do you point the module at a folder other than the default `input` folder?

Thanks!

Oh, and if this isn't the best forum for these kinds of questions do let me know and I'll redirect them to wherever is best!

**Comment by daveaglick** [comment]
You're _very_ close!

Deployment pipelines have a few special qualities:
- They only run when the `deploy` command is called (`dotnet run -- deploy`).
- They can read the output of other pipelines in their `Output` phase.
- Their `Output` phases always happen last.

This means that any modules inside the `Output` phase of a deployment pipeline get a "last pass" at anything other pipelines generated or output.

> Is it a case of using the CopyFiles module as an OutputModule?

`CopyFiles` is intended for copying from the `input` folder to the `output` folder. What you're really looking for during deployment is taking files from the `output` folder and uploading them somewhere else like a host.

> If so, how do you point the module at a folder other than the default input folder?

There are a few canned modules geared specifically for deployment (with more planned):

- `DeployNetlifySite` (in the `Statiq.Web.Netlify` package)
- `DeployGitHubPages` (in the `Statiq.Web.GitHub` package)
- `DeployAppService` (in the `Statiq.Web.Azure` package)

These modules are generally placed in the `Output` phase of a deployment pipeline. They all get the output files using `IExecutionContext.FileSystem.GetOutputPath()` and then either iterate the files there using normal `System.IO` calls or the built-in file system. That same strategy works well in your own deployment module if you're able to write deployment code.

Otherwise, if your deployment is typically done via the command line, you can also use a `StartProcess` module in the `Output` phase of a deployment pipeline. That gives you a lot of control over forking a process to handle deployment and you can pass it the output path obtained from `IExecutionContext.FileSystem.GetOutputPath()`.
